streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -80,7 +80,13 @@ void closeTask(final StreamTask task, final boolean clean) {
         }
     }
 
+    // Note that restoredPartitions may be nonempty even though restoring is, and will not be cleared in
+    // #updateRestored if this returns false, so we should clear any remaining partitions if restoring is empty
     boolean hasRestoringTasks() {
+        if (restoring.isEmpty()) {
```

Comment:
I felt it is a bit anti-pattern to modify states in a `hasXXX` function which is usually supposed to be side-effects free.

And by reading the code it seems every place where we may modify `restoring` or `running` we've also clear `restoredPartitions` and `restoringPartitions` already, so is this part still necessary?

Reply:
Agreed, rewrote this functionality so it's not a hidden side effect but an explicit call.

Regarding the "why", we want to make sure that `restoredPartitions` is cleared once everything is in running. Previously, all tasks followed the created -> restoring -> running transition after a rebalance, so we would always go through `AssignedStreamTasks#updateRestored` for the thread to transition from PARTITIONS_ASSIGNED to RUNNING, at which point we clear the (non-empty) `restoredPartitions`.
But now we may go from having some restoring tasks to having only running tasks just during the course of a rebalance, and would remove only the revoked tasks from `restoredPartitions` -- the partitions of the currently running tasks would still be in `restoredPartitions`, but if there are no other restoring tasks we would never hit `updateRestored` and thus not clear `restoredPartitions`